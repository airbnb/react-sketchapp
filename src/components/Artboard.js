// @flow
import * as React from 'react';
import * as PropTypes from 'prop-types';
import { or } from 'airbnb-prop-types';
import StyleSheet from '../stylesheet';
import ViewStylePropTypes from './ViewStylePropTypes';

const ViewportPropTypes = {
  name: PropTypes.string,
  width: PropTypes.number,
  height: PropTypes.number,
};

// $FlowFixMe
export default class Artboard extends React.Component {
  static propTypes = {
    // TODO(lmr): do some nice warning stuff like RN does
    style: or([PropTypes.shape(ViewStylePropTypes), PropTypes.number]),
    name: PropTypes.string,
    children: PropTypes.node,
    viewport: PropTypes.shape(ViewportPropTypes),
  };

  static defaultProps = {
    name: 'Artboard',
  };

  render() {
    return (
      <artboard
        style={StyleSheet.flatten(this.props.style)}
        name={this.props.name}
        viewport={this.props.viewport}
      >
        {this.props.children}
      </artboard>
    );
  }
}
