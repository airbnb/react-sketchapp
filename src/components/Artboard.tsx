import * as React from 'react';
import * as PropTypes from 'prop-types';
import { or } from 'airbnb-prop-types';
import StyleSheet from '../stylesheet';
import ViewStylePropTypes from './ViewStylePropTypes';

const ViewportPropTypes = {
  name: PropTypes.string,
  width: PropTypes.number,
  height: PropTypes.number,
};

export const ArtboardPropTypes = {
  style: or([PropTypes.shape(ViewStylePropTypes), PropTypes.number]),
  name: PropTypes.string,
  isHome: PropTypes.bool,
  children: PropTypes.node,
  viewport: PropTypes.shape(ViewportPropTypes),
};

export type Props = PropTypes.InferProps<typeof ArtboardPropTypes>;

export default class Artboard extends React.Component<Props> {
  static propTypes = ArtboardPropTypes;

  static defaultProps = {
    name: 'Artboard',
  };

  render() {
    return (
      <sketch_artboard
        style={StyleSheet.flatten(this.props.style)}
        name={this.props.name}
        viewport={this.props.viewport}
        isHome={this.props.isHome}
      >
        {this.props.children}
      </sketch_artboard>
    );
  }
}
