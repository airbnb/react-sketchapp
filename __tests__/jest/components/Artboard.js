import * as React from 'react';
import * as renderer from 'react-test-renderer';
import Artboard from '../../../src/components/Artboard';
import StyleSheet from '../../../src/stylesheet';

describe('<Artboard />', () => {
  it('renders children', () => {
    const tree = renderer
      .create(
        <Artboard>
          <foo>
            <bar />
          </foo>
        </Artboard>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it.todo('flattens its stylesheet');

  describe('name', () => {
    it('passes its name', () => {
      const tree = renderer.create(<Artboard name="bar" />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('defaults to Artboard', () => {
      const tree = renderer.create(<Artboard />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('style', () => {
    const styles = StyleSheet.create({
      view: {
        flex: 1,
      },
    });

    it('accepts a plain object', () => {
      const tree = renderer.create(<Artboard style={{ flex: 1 }} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('accepts a StyleSheet ordinal', () => {
      const tree = renderer.create(<Artboard style={styles.view} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('accepts an array of plain objects and/or StyleSheet ordinals', () => {
      const tree = renderer.create(<Artboard style={[{ flexGrow: 1 }, styles.view]} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('accepts artboard viewport preset', () => {
      const tree = renderer
        .create(
          <Artboard
            style={{ flex: 1, width: 360, height: 1280 }}
            viewport={{
              name: 'Mobile',
              width: 360,
              height: 640,
            }}
          />,
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
