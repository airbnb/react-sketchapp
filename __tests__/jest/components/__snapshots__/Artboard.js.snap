// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Artboard /> name defaults to Artboard 1`] = `
<artboard
  name="Artboard"
/>
`;

exports[`<Artboard /> name passes its name 1`] = `
<artboard
  name="bar"
/>
`;

exports[`<Artboard /> renders children 1`] = `
<artboard
  name="Artboard"
>
  <foo>
    <bar />
  </foo>
</artboard>
`;

exports[`<Artboard /> style accepts a StyleSheet ordinal 1`] = `
<artboard
  name="Artboard"
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`<Artboard /> style accepts a plain object 1`] = `
<artboard
  name="Artboard"
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`<Artboard /> style accepts an array of plain objects and/or StyleSheet ordinals 1`] = `
<artboard
  name="Artboard"
  style={
    Object {
      "flex": 1,
      "flexGrow": 1,
    }
  }
/>
`;

exports[`<Artboard /> style accepts artboard viewport preset 1`] = `
<artboard
  name="Artboard"
  style={
    Object {
      "flex": 1,
      "height": 1280,
      "width": 360,
    }
  }
  viewport={
    Object {
      "height": 640,
      "name": "Mobile",
      "width": 360,
    }
  }
/>
`;
